app/src/main/java/org/mozilla/focus/widget/ProgressBar.kt

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.widget
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.view.View
+import android.arch.lifecycle.Observer
+import android.os.Handler
+import android.os.Looper
+import android.os.Message
+import android.widget.LinearLayout
+import android.widget.TextView
+import org.mozilla.focus.R
+import org.mozilla.focus.architecture.NonNullObserver
+import org.mozilla.focus.fragment.BrowserFragment
+import org.mozilla.focus.utils.GifImageView
+import java.lang.ref.WeakReference
+import java.util.concurrent.TimeUnit
+
+private const val HIDE_MESSAGE_ID = 0
+private const val HIDE_ANIMATION_DURATION_MILLIS = 250L
+private val HIDE_AFTER_MILLIS = TimeUnit.SECONDS.toMillis(3)
+
+class ProgressBar @JvmOverloads constructor(
+        context: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
+    : LinearLayout(context, attrs, defStyle) {
+
+    private val hideHandler = ProgressBarHideHandler(this)
+    private var browserFrag: BrowserFragment? = null
+
+    fun setBrowserFragment(browserFragment: BrowserFragment) {
+        this.browserFrag = browserFragment
+        val textView = findViewById<TextView>(R.id.url)
+
+        browserFrag!!.session.url.observe(browserFrag!!, Observer { url -> textView.text = url })
+        browserFrag!!.session.loading.observe(browserFrag!!, object : NonNullObserver<Boolean>() {
+            public override fun onValueChanged(loading: Boolean) {
+                if (loading) {
+                    startUpdates()
+                }
+            }
+        })
+    }
+
+    init {
+        LayoutInflater.from(context)
+                .inflate(R.layout.progress_bar, this, true)
+        val gifImageView = findViewById<View>(R.id.gifImageView) as GifImageView
+        gifImageView.setGifImageResource(R.drawable.progress_loading_indicator)
+    }
+
+    fun startUpdates() {
+        setMaxVisibility()
+        resetCountdown()
+    }
+
+    private fun setMaxVisibility() {
+        animate().cancel()
+        alpha = 1f
+    }
+
+    private fun resetCountdown() {
+        hideHandler.removeMessages(HIDE_MESSAGE_ID)
+        hideHandler.sendEmptyMessageDelayed(HIDE_MESSAGE_ID, HIDE_AFTER_MILLIS)
```

Comment:
We want to hide the progress bar after the page is done loading, not after HIDE_AFTER_MILLIS, right?

Reply:
The reason why we used the Cursor implementation of disappearing was because the listener to the loading isn't reliable as it loads and "finishes loading" every time the url changes (and the url changes quite frequently depending on the entered url, e.g. www.youtube.com --> https://www.youtube.com --> https://www.youtube.com/#tv or smth)

When I just relied on the listener, it faded in and out every time the url changed, so chenxia led me to the cursor implementation on how to handle this case. Thoughts?

Reply:
@bkmunar The countdown suggestion is for solving the problem of the loads/redirects. However, I don't think starting this countdown every time you start the load makes sense (what if the network is really slow?). You should still hook this into the "done loading" cycle.